kinto/views/records.py

```diff
@@ -55,23 +53,32 @@ def process_record(self, new, old=None):
         if not schema or not asbool(settings.get(schema_validation)):
             return new
 
-        collection_timestamp = self._collection[self.model.modified_field]
-
+        # Remove internal and auto-assigned fields from schema and record.
+        internal_fields = (self.model.id_field,
+                           self.model.modified_field,
+                           self.schema_field,
+                           self.model.permissions_field)
+        required_fields = [f for f in schema.get('required', []) if f not in internal_fields]
+        if required_fields:
+            schema = {**schema, 'required': required_fields}
+        else:
+            schema = {f: v for f, v in new.items() if f != 'required'}
```

Comment:
I don't think we need to go to this lengths to avoid putting an empty `required` field. The [JSON Schema Spec](http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.17) says that omitting it is the same as an empty array. Unless you're trying to do this so that we don't confuse users by secretly changing their schema, but I don't think that's valuable, since we're already changing the schema.

Should we be doing these manipulations to the schema at schema save time?

Should we be removing definitions of internal fields from `schema['properties']`?